# Phase 1 — Data Understanding & Profiling

# Setup

Importing libraries

In [ ]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 120)

# 2. Table viewing and basic info
Viewing the first 5 rows of the table along with count total of rows and columns

In [11]:
year_2009_2010 = pd.read_excel("../data/raw/online_retail_II.xlsx", sheet_name="Year 2009-2010")
print(f"year_2009_2010: {year_2009_2010.shape[0]} rows, {year_2009_2010.shape[1]} columns")
display(year_2009_2010.head(5))

year_2009_2010: 525461 rows, 8 columns


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [12]:
year_2010_2011 = pd.read_excel("../data/raw/online_retail_II.xlsx", sheet_name="Year 2010-2011")
print(f"year_2010_2011: {year_2010_2011.shape[0]} rows, {year_2010_2011.shape[1]} columns")
display(year_2010_2011.head(5))

year_2010_2011: 541910 rows, 8 columns


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [13]:
# Inspecting the data types and null values
print(f"table: retail_2009_2010")
print(year_2009_2010.info())

table: retail_2009_2010
<class 'pandas.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[us]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 32.1+ MB
None


In [14]:
# Inspecting the data types and null values
print(f"table: retail_2010_2011")
print(year_2010_2011.info())

table: retail_2010_2011
<class 'pandas.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1   StockCode    541910 non-null  object        
 2   Description  540456 non-null  object        
 3   Quantity     541910 non-null  int64         
 4   InvoiceDate  541910 non-null  datetime64[us]
 5   Price        541910 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      541910 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB
None


# 3. Missing Audit

In [15]:
print(f"table: year_2009_2010")
print(year_2009_2010.isnull().sum())

table: year_2009_2010
Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64


In [22]:
print(f"table: year_2009_2010")
null_values_y1 = year_2009_2010.isnull().mean() * 100
print(null_values_y1[null_values_y1 > 0].sort_values(ascending=False))

table: year_2009_2010
Customer ID    20.539488
Description     0.557225
dtype: float64


In [50]:
year_2009_2010[year_2009_2010["Description"].isnull() & year_2009_2010["Customer ID"].isnull()]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
3161,489659,21350,NaN,230,2009-12-01 17:39:00,0.0,NaN,United Kingdom
3731,489781,84292,NaN,17,2009-12-02 11:45:00,0.0,NaN,United Kingdom
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
524730,538141,22218,NaN,2,2010-12-09 15:58:00,0.0,NaN,United Kingdom
524731,538143,84534B,NaN,1,2010-12-09 15:58:00,0.0,NaN,United Kingdom
525231,538159,21324,NaN,-18,2010-12-09 17:17:00,0.0,NaN,United Kingdom
525232,538158,20892,NaN,-32,2010-12-09 17:17:00,0.0,NaN,United Kingdom


In [21]:
print(f"table: year_2010_2011")
print(year_2010_2011.isnull().sum())

table: year_2010_2011
Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64


In [23]:
print(f"table: year_2010_2011")
null_values_y2 = year_2010_2011.isnull().mean() * 100
print(null_values_y2[null_values_y2 > 0].sort_values(ascending=False))

table: year_2010_2011
Customer ID    24.926648
Description     0.268310
dtype: float64


In [51]:
year_2010_2011[year_2010_2011["Description"].isnull() & year_2010_2011["Customer ID"].isnull()]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1510,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1985,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1986,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
2022,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,2011-12-07 18:26:00,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom


# 4. Duplicate checks

In [52]:
print(f"table: year_2009_2010")
print(f"duplicates: {year_2009_2010.duplicated().sum()}")
print(f"{year_2009_2010.duplicated().mean() * 100:.2f}% of rows are duplicates")

print(f"table: year_2010_2011")
print(f"duplicates: {year_2010_2011.duplicated().sum()}")
print(f"{year_2010_2011.duplicated().mean() * 100:.2f}% of rows are duplicates")

table: year_2009_2010
duplicates: 6865
1.31% of rows are duplicates
table: year_2010_2011
duplicates: 5268
0.97% of rows are duplicates


## 5. Categorical value inspection
Checking the distribution of categorical variables

In [32]:
year_2009_2010["Country"].value_counts()

Country
United Kingdom          485852
EIRE                      9670
Germany                   8129
France                    5772
Netherlands               2769
Spain                     1278
Switzerland               1187
Portugal                  1101
Belgium                   1054
Channel Islands            906
Sweden                     902
Italy                      731
Australia                  654
Cyprus                     554
Austria                    537
Greece                     517
United Arab Emirates       432
Denmark                    428
Norway                     369
Finland                    354
Unspecified                310
USA                        244
Japan                      224
Poland                     194
Malta                      172
Lithuania                  154
Singapore                  117
RSA                        111
Bahrain                    107
Canada                      77
Hong Kong                   76
Thailand                    76


In [33]:
year_2010_2011["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8558
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


From the categorical value inspection it is found that, we are getting maximum customers from UK, so the data would naturally be more inclined towards UK.

## 6. Numerical value check
Descriptive statistics for numerical columns

In [34]:
year_2009_2010.describe(exclude= "object")

,Quantity,InvoiceDate,Price,Customer ID
count,525461.000000,525461,525461.000000,417534.000000
mean,10.337667,2010-06-28 11:37:36.845018,4.688834,15360.645478
min,-9600.000000,2009-12-01 07:45:00,-53594.360000,12346.000000
25%,1.000000,2010-03-21 12:20:00,1.250000,13983.000000
50%,3.000000,2010-07-06 09:51:00,2.100000,15311.000000
75%,10.000000,2010-10-15 12:45:00,4.210000,16799.000000
max,19152.000000,2010-12-09 20:01:00,25111.090000,18287.000000
std,107.424110,NaN,146.126914,1680.811316


In [35]:
year_2010_2011.describe(exclude= "object")

,Quantity,InvoiceDate,Price,Customer ID
count,541910.000000,541910,541910.000000,406830.000000
mean,9.552234,2011-07-04 13:35:22.342307,4.611138,15287.684160
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.080957,NaN,96.759765,1713.603074


# 7. Profiling of few columns

In [44]:
print("table: year_2009_2010")
print(f"start date : {year_2009_2010["InvoiceDate"].min()}")
print(f"end date : {year_2009_2010["InvoiceDate"].max()}")

print("------------------------------")

print("table: year_2010_2011")
print(f"start date : {year_2010_2011["InvoiceDate"].min()}")
print(f"end date : {year_2010_2011["InvoiceDate"].max()}")

table: year_2009_2010
start date : 2009-12-01 07:45:00
end date : 2010-12-09 20:01:00
------------------------------
table: year_2010_2011
start date : 2010-12-01 08:26:00
end date : 2011-12-09 12:50:00


For the year 2009-2010 the date ranges from 01-12-2009 to 09-12-2010

For the year 2009-2010 the date ranges from 01-12-2010 to 09-12-2010

In [ ]:
print(f"total customers for the year 2009-2010: {year_2009_2010['Customer ID'].nunique()}")

print(f"total customers for the year 2010-2011: {year_2010_2011['Customer ID'].nunique()}")

unique customers for the year 2009-2010: 4383
unique customers for the year 2010-2011: 4372


In [47]:
print(f"countries for the year 2009-2010: {year_2009_2010['Country'].unique()}")

print(f"countries for the year 2010-2011: {year_2010_2011['Country'].unique()}")

countries for the year 2009-2010: <StringArray>
[      'United Kingdom',               'France',                  'USA',              'Belgium',
            'Australia',                 'EIRE',              'Germany',             'Portugal',
                'Japan',              'Denmark',              'Nigeria',          'Netherlands',
               'Poland',                'Spain',      'Channel Islands',                'Italy',
               'Cyprus',               'Greece',               'Norway',              'Austria',
               'Sweden', 'United Arab Emirates',              'Finland',          'Switzerland',
          'Unspecified',                'Malta',              'Bahrain',                  'RSA',
              'Bermuda',            'Hong Kong',            'Singapore',             'Thailand',
               'Israel',            'Lithuania',          'West Indies',              'Lebanon',
                'Korea',               'Brazil',               'Canada',       

In [48]:
print(f"distinct products sold in the year 2009-2010: {year_2009_2010['Description'].nunique()}")

print(f"distinct products sold in the year 2010-2011: {year_2010_2011['Description'].nunique()}")

distinct products sold in the year 2009-2010: 4681
distinct products sold in the year 2010-2011: 4223


We have a total of 4383 customers shopping 4681 distinct products from us across 40 different countries in the year 2009-2010

We have a total of 4372 customers shopping 4223 distinct products from us across 38 different countries in the year 2010-2011